test/Npgsql.Tests/Types/CompositeHandlerTests.cs

```diff
@@ -0,0 +1,141 @@
+            var name = nameTranslator.TranslateTypeName(typeof(T).Name + nameSuffix);
+
+            if (schema == null)
+                nameQualified = name;
+            else
+            {
+                schema = nameTranslator.TranslateTypeName(schema);
+                nameQualified = schema + "." + name;
+            }
+
+            var connection = OpenConnection();
+
+            try
+            {
+                connection.ExecuteNonQuery(schema is null? $"DROP TYPE IF EXISTS {name}" : $"DROP SCHEMA IF EXISTS {schema} CASCADE; CREATE SCHEMA {schema}");
+                connection.ExecuteNonQuery($"CREATE TYPE {nameQualified} AS ({composite.GetAttributes()})");
+
+                connection.ReloadTypes();
+                connection.TypeMapper.MapComposite<T>(nameQualified, nameTranslator);
+
+                return connection;
+            }
+            catch
+            {
+                connection.Dispose();
+                throw;
+            }
+        }
+
+        interface IComposite
+        {
+            string GetAttributes();
+            string GetValues();
+        }
+
+        public class ClassWithProperty : IComposite
+        {
+            public string? Value { get; set; }
+            public string GetAttributes() => "value text";
+	        public string GetValues() => $"'{Value}'";
+        }
+
+        public class ClassWithField : IComposite
+        {
+            public string? Value;
+            public string GetAttributes() => "value text";
+	        public string GetValues() => $"'{Value}'";
+        }
+
+        public struct StructWithProperty : IComposite
+        {
+            public string Value { get; set; }
+            public string GetAttributes() => "value text";
+	        public string GetValues() => $"'{Value}'";
+        }
+
+        public struct StructWithField : IComposite
+        {
+            public string Value;
+            public string GetAttributes() => "value text";
+	        public string GetValues() => $"'{Value}'";
+        }
+
+        public class TypeWithTwoProperties : IComposite
+        {
+            public string GetAttributes() => "int_value integer, string_value text";
+            public string GetValues() => $"{IntValue}, '{StringValue}'";
+
+            public int IntValue { get; set; }
+            public string? StringValue { get; set; }
+        }
+
+        public class TypeWithTwoPropertiesInverted : IComposite
+        {
+            public string GetAttributes() => "string_value text, int_value integer";
+            public string GetValues() => $"'{StringValue}', {IntValue}";
+
+            public int IntValue { get; set; }
+            public string? StringValue { get; set; }
+        }
+
+        public abstract class SimpleComposite : IComposite
+        {
+            public string Value { get; set; } = "hello";
+            public string GetAttributes() => "value text";
+	        public string GetValues() => $"'{Value}'";
```

Comment:
Looks like you reviewed it for a very long time since it's outdated before a fix. But the wrong indentation is still present in other lines due to tabs.

Reply:
Yes, there are comments from an old previous review.